subprojects/language-java/src/main/java/org/gradle/api/tasks/JavaExec.java

```diff
@@ -131,18 +130,37 @@ public JavaExec() {
         javaExecSpec.getMainClass().convention(mainClass);
         javaExecSpec.getMainModule().convention(mainModule);
         javaExecSpec.getModularity().getInferModulePath().convention(modularity.getInferModulePath());
-        javaLauncher = objectFactory.property(JavaLauncher.class);
+
+        JavaToolchainService javaToolchainService = getJavaToolchainService();
+        Provider<JavaLauncher> javaLauncherConvention = getProviderFactory()
+            .provider(() -> JavaExecExecutableUtils.getExecutableOverrideToolchainSpec(this, objectFactory))
+            .flatMap(javaToolchainService::launcherFor)
+            .orElse(javaToolchainService.launcherFor(it -> {}));
+        javaLauncher = objectFactory.property(JavaLauncher.class).convention(javaLauncherConvention);
+        javaLauncher.finalizeValueOnRead();
     }
 
     @TaskAction
     public void exec() {
+        validateExecutableMatchesToolchain();
         setJvmArgs(getJvmArgs()); // convention mapping for 'jvmArgs'
         JavaExecAction javaExecAction = getExecActionFactory().newJavaExecAction();
         javaExecSpec.copyTo(javaExecAction);
-        javaExecAction.setExecutable(getEffectiveExecutable());
+        String effectiveExecutable = getJavaLauncher().get().getExecutablePath().toString();
+        javaExecAction.setExecutable(effectiveExecutable);
+
         execResult.set(javaExecAction.execute());
     }
 
+    private void validateExecutableMatchesToolchain() {
+        File toolchainExecutable = getJavaLauncher().get().getExecutablePath().getAsFile();
+        String customExecutable = getExecutable();
+        checkState(
+            customExecutable == null || new File(customExecutable).equals(toolchainExecutable),
+            "Toolchain from `executable` property does not match toolchain from `javaLauncher` property"
```

Comment:
This message might be more helpful if it included the task name containing the problem - this would point the author directly to the task that needed to be fixed in a large build.  It also might be better to throw some sort of custom Gradle exception rather than `ISE`.

Reply:
When task fails at execution time, we already capture the context in which the failure occurs and include it when printing on the console:
```
* What went wrong:
Execution failed for task ':run'.
> Toolchain from `executable` property does not match toolchain from `javaLauncher` property
```

Having said that, we should improve the tests that expect this failure and check that the error actually happens under the given task. I'll fix this shortly.

Reply:
I didn't realize this would be an execution time failure and print the name.  Adding that expectation to the tests is a good idea.